pandas/core/internals/managers.py

```diff
@@ -235,16 +235,19 @@ def shape(self) -> Shape:
     def ndim(self) -> int:
         return len(self.axes)
 
-    def set_axis(self, axis: int, new_labels: Index) -> None:
+    def set_axis(
+        self, axis: int, new_labels: Index, verify_integrity: bool = True
```

Comment:
I added a `verify_integrity` keyword here, and put the length verification behind `if verify_integrity`. Reason I needed this is because in groupby, we are sometimes setting an Index with a different length as the original one.